sound/soc/sof/ipc4-telemetry.c

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2018-2023 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <linux/io.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/debug.h>
+#include <sound/sof/ipc4/header.h>
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-telemetry.h"
+#include "ipc4-priv.h"
+
+static void __iomem *sof_ipc4_query_exception_address(struct snd_sof_dev *sdev)
+{
+	u32 type = SOF_IPC4_DEBUG_SLOT_TELEMETRY;
+	u32 telemetry_slot_offset;
+	void __iomem *io_addr;
+	u32 offset;
+
+	telemetry_slot_offset = sof_ipc4_find_debug_slot_offset_by_type(sdev, type);
+	if (!telemetry_slot_offset)
+		return NULL;
+
+	/* skip the first separator magic number */
+	offset = telemetry_slot_offset + sizeof(u32);
+	io_addr = sdev->bar[sdev->mailbox_bar] + offset;
```

Comment:
@ujfalusi  We have a tool for coredump in zephyr: coredump_gdbserver.py which doesn't recognize the magic separator number, so we need to strip it.

Reply:
@RanderWang, and if the dump sub-slot is moved and not at the beginning of the telemetry slot?
What I mean is that via this debugfs file we can grab the _whole_ telemetry slot and not only the panic dump section.
The dump is just part of that slot, right?
If we say that we offer the whole slot then we should offer the whole slot, if we say that we only offer the panic information, then we should do only  that, and not whole slot - sizeof(u32)

Reply:
@ujfalusi At first we don't support panic dump with debugfs. According to @kv2019i , we need to support exception usage defined by IPC3 which is familiar with user, also including some debug tools, so we develop the debugfs dump according to IPC3 usage. @kv2019i  could you help to make it more clear ? Thanks!

Reply:
@RanderWang, I just want to avoid us changing the definition what user should expect from the debugfs file.
We are creating the file named as 'exception' so it should be the exception part of the telemetry and only that, so the `SOF_IPC4_DEBUG_SLOT_SIZE - 4` as length to copy to user is not correct, or is it? I don't think the panic dump takes up the whole telemetry slot...

Reply:
@ujfalusi @RanderWang  As per current definition, it does only contain the exception and the exception tooling (which is standard Zephyr) doesn't understand the 32bit header, so if kernel exposes that, we'd need another user-space step to strip that. That's not a big deal of course, but given the logical name is "exception", to me it would make sense to strip the 32bit header as that is just SOF way to find the exception from the telemetry slot.

If in later ABI change, the telemetry slot is extended with a new 32bit identifier that basicly tells the kernel driver that now the debug slot has more data than just the exception, kernel would need to parse that and expose the individual pieces via different interfaces (and only exposes the exception via "exception"). But then we'd break existing user-space that expects to get a magic 32bit at start of exception. So this is not good, I'd rather have the "exception" debugfs file match the contents of standard Zephyr exception and be able to feed that to the tools directly.